Update how towns/dungeons/gyms/routes requirements are handled

RedSparr0w opened the pull request:
All requirements should be updated,

Note:
Some will need more requirements added, as they weren't available before,
not sure which though, so will need to be done in a separate PR.

Edit:
Kanto requirements are all updated, some routes/dungeons/gyms are optional until specific points in the game.
Also had to update the Tutorial Quests as Viridian Forest is required before you reach Brock

RedSparr0w committed:
Add all requirements as 1 array for towns/dungeons

RedSparr0w committed:
Fixup dungeon name in notification

RedSparr0w committed:
Don't alert if requirement already completed

RedSparr0w committed:
Update how dungeons are handled

RedSparr0w committed:
Remove the requirements from starting dungeon

Shouldn't be able to access the dungeon to start it anyway

RedSparr0w committed:
Update Gym requirements

RedSparr0w committed:
Merge branch 'develop' into refactor/requirements

RedSparr0w committed:
Update Champion, GymBattle

RedSparr0w committed:
Merge branch 'refactor/requirements' of https://github.com/pokeclicker-dev/pokeclicker into refactor/requirements

RedSparr0w committed:
Refactor Kanto requirements/routes

RedSparr0w committed:
Normalize route numbers HP/rewards etc

RedSparr0w committed:
Add route 6 as requirement (in place of SS Anne) for cut

RedSparr0w committed:
Update Johto routes

RedSparr0w committed:
Update Hoenn Routes

RedSparr0w committed:
Fixup things that still needed fixing

RedSparr0w committed:
Also this :sweat:

Aegyo commented on src/scripts/worldmap/Routes.ts:
```diff
@@ -0,0 +1,777 @@
+    public static normalizedNumber(region: GameConstants.Region, route: number): number {
+        return this.regionRoutes
+            .sort((routeA, routeB) => routeA.orderNumber - routeB.orderNumber)
+            .sort((routeA, routeB) => routeA.region - routeB.region)
```
Seems like something that will be running a lot, and return the same thing every time. Why don't we just do the sorting once after all routes have been added?

RedSparr0w replied:
Good idea,
I've updated it to sort when a new route is added.

Aegyo commented on src/scripts/worldmap/MapHelper.ts:
```diff
@@ -205,40 +156,21 @@ class MapHelper {
+            town.requirements.forEach(requirement => {
+                if (requirement.isCompleted()) {
+                    return;
+                } else if (requirement instanceof GymBadgeRequirement) {
```
Not liking these `instanceof` checks which will need adding to if/when we add new Requirement types. Maybe Requirement should have a required `hint(): string` method we can call, and new requirements need only implement it?

RedSparr0w replied:
Done!
Not sure if this is correct though, it doesn't seem to be required as an abstract.
Is there a better way?

Aegyo reviewed (commented):
Haven't had a chance to play through the game on this branch yet, but here are the only issues I have after browsing all of the changes.

There are few requirements for things which seem redundant (eg, route 4 cleared and champion defeated for cerulean cave - route 4 cleared seems necessary to reach the champion), but it doesn't particularly bother me.

Also `EitherRequirement` suggests only two, but it can be many. Not sure if you want to bother renaming it to match `Array.some` or `Promise.any` (and a possible future `every`/`all`)

RedSparr0w committed:
Add hint text to each of the requirements

RedSparr0w committed:
Route -> RegionRoute

RedSparr0w committed:
EitherRequirement -> OneFromManyRequirement

RedSparr0w committed:
remove some old requirements

RedSparr0w commented:
Changed from `EitherRequirement` -> `OneFromManyRequirement` not sure if that makes more sense, not sure what else to call it though.

Also removed some of the redundant requirements.

Aegyo reviewed (commented):
After a quick play through, had a few thoughts:

When trying to access Cinnabar gym it tells me I am not worthy and should try completing more badges first, but the real reason I can't access is because I haven't done Pokemon Mansion. Can we update the gym access hint to use the Requirements for that gym? I could just do all of the orange things on the map to unlock it, but it would be nice to be told the exact reason I can't enter the gym.

Routes 47/48 can be accessed after Cianwood, but don't have an orderNumber to reflect this - is that on purpose?

Kanto had surf areas unlocked from both ends when the badge was unlocked, but Hoenn doesn't seem to do this. I would add route 109, 134, and New Mauville to the things unlocked by Petalburg gym, and also have 108 etc be opened from either side. 134 should not give access to 133 though, there are strong currents preventing that in the actual games.

Route 116 could do with an orderNumber, and possibly the water routes 105-109

RedSparr0w commented:
I haven't done any of the requirement changes for Johto and Hoenn yet, only refactored them for the new format using the previous data.

> Can we update the gym access hint to use the Requirements for that gym? I could just do all of the orange things on the map to unlock it, but it would be nice to be told the exact reason I can't enter the gym.

Good idea, I completely forgot about those!

RedSparr0w committed:
Merge branch 'develop' into refactor/requirements

Aegyo commented:
Ah okay, are the other region requirements going to be left for another PR, so that this can be merged?

RedSparr0w committed:
Add check in case no requirements

RedSparr0w commented:
Yeah might just update the other regions in the Sinnoh branch when I adjust that to fit this PR too.

RedSparr0w committed:
Add reason Gym cannot be accessed yet

RedSparr0w committed:
Only show town is incomplete if Gym is accessable and badge not obtained

RedSparr0w commented:
Sweet, this should be ready for another review.

I've updated the Gyms to give the reason they cannot be accessed yet.

Also made it so the town won't show as orange if the Gym cannot be accessed yet.

Attached save pre Cinnabar Pokemon Mansion for testing:

[Pre Cinnabar Mansion.txt](https://github.com/pokeclicker-dev/pokeclicker/files/5074480/Pre.Cinnabar.Mansion.txt)

Aegyo reviewed (approved):
Looks good :+1:

RedSparr0w committed:
Merge branch 'develop' into refactor/requirements
